src/frontend/src/handler/create_source.rs

```diff
@@ -515,32 +517,64 @@ pub fn handle_addition_columns(
-    };
+    let gen_default_column_name =
+        |connector_name: &str, addi_column_name: &str, inner_field: Option<&String>| {
+            if let Some(inner) = inner_field {
+                format!("_rw_{}_{}_{}", connector_name, addi_column_name, inner)
+            } else {
+                format!("_rw_{}_{}", connector_name, addi_column_name)
+            }
+        };
 
     let latest_col_id: ColumnId = columns
         .iter()
         .map(|col| col.column_desc.column_id)
         .max()
         .unwrap(); // there must be at least one column in the column catalog
 
-    for (col_name, gen_column_catalog_fn) in addition_col_list {
-        // always insert in spec order
-        if let Some(idx) = additional_columns
-            .iter()
-            .position(|(col, _)| col.real_value().eq_ignore_ascii_case(col_name))
+    while let Some(item) = additional_columns.pop() {
         {
-            let (_, alias) = additional_columns.remove(idx);
-            columns.push(gen_column_catalog_fn(
-                latest_col_id.next(),
-                alias
-                    .map(|alias| alias.real_value())
-                    .unwrap_or_else(|| gen_default_column_name(connector_name.as_str(), col_name))
-                    .as_str(),
-            ))
+            // only allow header column have inner field
+            if item.inner_field.is_some()
+                && !item.column_type.real_value().eq_ignore_ascii_case("header")
+            {
+                return Err(RwError::from(ProtocolError(format!(
+                    "Only header column can have inner field, but got {:?}",
+                    item.column_type.real_value(),
+                ))));
+            }
         }
+
+        let (_, gen_column_catalog_fn) = addition_col_list
+            // todo: refactor `addition_col_hashmap` to HashMap<&str, CompatibleAdditionalColumnsFn>
+            .iter()
+            .find(|ele| {
+                ele.0
+                    .eq_ignore_ascii_case(item.column_type.real_value().as_str())
+            })
+            .ok_or_else(|| {
+                RwError::from(ProtocolError(format!(
+                    "Unknown additional column type {:?}",
+                    item.column_type.real_value()
+                )))
+            })?;
+
+        columns.push(gen_column_catalog_fn(
+            latest_col_id.next(),
+            item.column_alias
+                .map(|alias| alias.real_value())
+                .unwrap_or_else(|| {
+                    gen_default_column_name(
+                        connector_name.as_str(),
+                        item.column_type.real_value().as_str(),
+                        item.inner_field.as_ref(),
+                    )
+                })
+                .as_str(),
+            item.inner_field.as_deref(),
+        ))
     }
+
     if !additional_columns.is_empty() {
         return Err(RwError::from(ProtocolError(format!(
             "Unknown additional columns {:?}",
```

Comment:
It seems that this branch will never be entered.

Reply:
You are right. Will remove it.